### PR TITLE
Etcd with credentials

### PR DESCRIPTION
**Pull request by GeorgianaElena**
Regarding https://github.com/jupyterhub/traefik-proxy/issues/18, I managed to enable TraefikEtcdProxy to authenticate with an etcd with credentials set up.

I had to change back to the synchronous etcd python client because with the async one we couldn't execute a transaction while authenticated (from my understanding, the client credentials aren't passed forward to the server during a transaction like it happens for the simple operations).

Also, I tried to run the etcd operations in a separate thread like in [PAMAuthenticator](https://github.com/jupyterhub/jupyterhub/blob/00cafc8392be5b32d0da2e869ddc74067ff27fc5/jupyterhub/auth.py).

@minrk, what do you think about this approach?

closes https://github.com/jupyterhub/traefik-proxy/issues/18

### CODE REVIEW

File: jupyterhub_traefik_proxy/etcd.py [hunk 1]
```diff
@@ -18,30 +18,44 @@
...
     def _default_client(self):
         etcd_service = urlparse(self.etcd_url)
-        return client(
-            endpoint=str(etcd_service.hostname) + ":" + str(etcd_service.port)
-        )
+        if self.etcd_password:
+            return etcd3.client(
+                host=str(etcd_service.hostname),
+                port=etcd_service.port,
+                user="root",
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: `etcd_username` should probably be a configurable trait like etcd_password. `root` can be the default, if desired.

**Comment by minrk** [comment]
This should be fine! I'm a bit surprised that the async client can't deal with credentials, but there's no problem at all running in a thread like this.

This should mean that requirements.txt needs updating to specify the new etcd client package and not the old one, right?

Can you update the tests to run etcd with a password to make sure everything is working?

**Comment by GeorgianaElena** [comment]
@minrk, I've already updated requirements.txt. The sync client is ``etcd3``.
Also, I added two new fixtures for testing the authentication (```auth_external_etcd_proxy``` and ```auth_etcd_proxy```).

You can see on travis that the tests are ran and pass for these new fixtures also.

File: jupyterhub_traefik_proxy/proxy.py [hunk 2]
```diff
@@ -143,7 +143,7 @@ def _launch_traefik(self, config_type):
                         "traefik",
                         "--etcd",
                         "--etcd.useapiv3=true",
-                        "--etcd.username=root",
+                        "--etcd.username=" + self.etcd_username,
                         "--etcd.password=" + self.etcd_password,
```
Comment (minrk) [inline jupyterhub_traefik_proxy/proxy.py]: Ah, I didn't catch this before. I think it's actually important to not set credentials on the command-line (other processes can see these values). Let's switch this to writing a private toml config file.

Comment (GeorgianaElena) [reply jupyterhub_traefik_proxy/proxy.py]: You're right @minrk. I've just added some changes addressing this.
```TraefikEtcdProxy``` will now load its static configuration (including the etcd credentials) from a toml file just like ```TraefikTomlProxy```.

File: jupyterhub_traefik_proxy/etcd.py [hunk 3]
```diff
@@ -59,52 +75,47 @@ def _default_client(self):
...
+        }
 
-        if status:
-            self.log.error(
-                "Couldn't set up traefik's static config. Response: %s", response
+        try:
+            traefik_utils.persist_static_conf(
+                self.toml_static_config_file, self.static_config
             )
+        except IOError:
```
Comment (minrk) [inline jupyterhub_traefik_proxy/etcd.py]: Let's always raise here. As a general rule, we only want to proceed past an error if it's recoverable. Failing to write this file should mean a hard exit, so letting it raise every time is probably what we want.

**Comment by minrk** [comment]
Great! I suggested one small change in error handling, but I think this should be all set otherwise.

File: jupyterhub_traefik_proxy/etcd.py (hunk 3 above)
Comment (GeorgianaElena) [reply jupyterhub_traefik_proxy/etcd.py]: You're right! I've just added this to the toml version also.

**Comment by minrk** [comment]
Great!

**Comment by minrk** [comment]
With this merged, I just marked 0.1.0 as a release, so we can drop the `--pre` argument when installing this. Great job!

**Pull request merged**